pkg/disk/disk.go

```diff
@@ -0,0 +1,254 @@
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	grpccodes "google.golang.org/grpc/codes"
+	grpcstatus "google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/longhorn/longhorn-spdk-engine/pkg/api"
+	spdkclient "github.com/longhorn/longhorn-spdk-engine/pkg/client"
+	spdkrpc "github.com/longhorn/longhorn-spdk-engine/proto/spdkrpc"
+
+	rpc "github.com/longhorn/longhorn-instance-manager/pkg/imrpc"
+	"github.com/longhorn/longhorn-instance-manager/pkg/meta"
+	"github.com/longhorn/longhorn-instance-manager/pkg/util"
+)
+
+const (
+	spdkTgtReadinessProbeTimeout = 60 * time.Second
+)
+
+type Server struct {
+	sync.RWMutex
+
+	spdkServiceAddress string
+	shutdownCh         chan error
+	HealthChecker      HealthChecker
+
+	spdkClient *spdkclient.SPDKClient
+}
+
+func NewServer(spdkEnabled bool, spdkServiceAddress string, shutdownCh chan error) (*Server, error) {
+	s := &Server{
+		spdkServiceAddress: spdkServiceAddress,
+		shutdownCh:         shutdownCh,
+		HealthChecker:      &GRPCHealthChecker{},
+	}
+
+	if spdkEnabled {
+		logrus.Info("Creating SPDK client since SPDK is enabled")
+
+		if !util.IsSPDKTgtReady(spdkTgtReadinessProbeTimeout) {
+			return nil, fmt.Errorf("spdk_tgt is not ready in %v", spdkTgtReadinessProbeTimeout)
+		}
+
+		spdkClient, err := spdkclient.NewSPDKClient(s.spdkServiceAddress)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create SPDK client")
+		}
+		s.spdkClient = spdkClient
+	}
+
+	go s.startMonitoring()
+
+	return s, nil
+}
+
+func (s *Server) startMonitoring() {
+	done := false
+	for {
+		select {
+		case <-s.shutdownCh:
+			logrus.Info("Disk gRPC Server is shutting down")
+			done = true
+		}
+		if done {
+			break
+		}
+	}
+}
+
+func (s *Server) VersionGet(ctx context.Context, req *empty.Empty) (*rpc.DiskVersionResponse, error) {
+	v := meta.GetDiskServiceVersion()
+	return &rpc.DiskVersionResponse{
+		Version:   v.Version,
+		GitCommit: v.GitCommit,
+		BuildDate: v.BuildDate,
+
+		InstanceManagerDiskServiceAPIVersion:    int64(v.InstanceManagerDiskServiceAPIVersion),
+		InstanceManagerDiskServiceAPIMinVersion: int64(v.InstanceManagerDiskServiceAPIMinVersion),
+	}, nil
+}
+
+func (s *Server) DiskCreate(ctx context.Context, req *rpc.DiskCreateRequest) (*rpc.Disk, error) {
+	log := logrus.WithFields(logrus.Fields{
+		"diskType":  req.DiskType,
+		"diskName":  req.DiskName,
+		"diskPath":  req.DiskPath,
+		"blockSize": req.BlockSize,
+	})
+
+	log.Info("Creating disk")
+
+	if req.DiskName == "" || req.DiskPath == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "disk name and disk path are required")
+	}
+
+	switch req.DiskType {
+	case rpc.DiskType_block:
+		ret, err := s.spdkClient.DiskCreate(req.DiskName, req.DiskPath, req.BlockSize)
+		if err != nil {
+			return nil, grpcstatus.Error(grpccodes.Internal, err.Error())
+		}
+		return spdkDiskToDisk(ret), nil
+	case rpc.DiskType_filesystem:
+		// TODO: implement filesystem disk type
```

Comment:
Let's create a ticket for this in the future.

Reply:
Tracked in https://github.com/longhorn/longhorn/issues/6024